drivers/i2c/i2c_cc23x0.c

```diff
@@ -0,0 +1,297 @@
+			return -EIO;
+		}
+
+		*buf = I2CControllerGetData(base);
+		return 0;
+	}
+
+	/* Burst receive */
+	I2CControllerCommand(base, I2C_CONTROLLER_CMD_BURST_RECEIVE_START);
+	k_sem_take(&data->complete, K_FOREVER);
+
+	if (data->error != I2C_MASTER_ERR_NONE) {
+		goto recv_error_stop;
+	}
+
+	buf[0] = I2CControllerGetData(base);
+
+	for (int i = 1; i < len - 1; i++) {
+		I2CControllerCommand(base, I2C_CONTROLLER_CMD_BURST_RECEIVE_CONT);
+		k_sem_take(&data->complete, K_FOREVER);
+
+		if (data->error != I2C_MASTER_ERR_NONE) {
+			goto recv_error_stop;
+		}
+
+		buf[i] = I2CControllerGetData(base);
+	}
+
+	I2CControllerCommand(base, I2C_CONTROLLER_CMD_BURST_RECEIVE_FINISH);
+	k_sem_take(&data->complete, K_FOREVER);
+
+	if (data->error != I2C_MASTER_ERR_NONE) {
+		return -EIO;
+	}
+
+	buf[len - 1] = I2CControllerGetData(base);
+
+	return 0;
+
+recv_error_stop:
+	I2CControllerCommand(base, I2C_CONTROLLER_CMD_BURST_RECEIVE_ERROR_STOP);
+	return -EIO;
+}
+
+static int i2c_cc23x0_transfer(const struct device *dev, struct i2c_msg *msgs, uint8_t num_msgs,
+			       uint16_t addr)
+{
+	struct i2c_cc23x0_data *data = dev->data;
+	int ret = 0;
+
+	if (num_msgs == 0) {
+		return 0;
+	}
+
+	k_sem_take(&data->lock, K_FOREVER);
+
+	for (int i = 0; i < num_msgs; i++) {
+		/* Not supported by hardware */
+		if (msgs[i].flags & I2C_MSG_ADDR_10_BITS) {
+			ret = -EIO;
+			break;
+		}
+
+		if ((msgs[i].flags & I2C_MSG_RW_MASK) == I2C_MSG_WRITE) {
+			ret = i2c_cc23x0_transmit(dev, msgs[i].buf, msgs[i].len, addr);
+		} else {
+			ret = i2c_cc23x0_receive(dev, msgs[i].buf, msgs[i].len, addr);
+		}
+
+		if (ret) {
+			break;
+		}
+	}
+	k_sem_give(&data->lock);
+
+	return ret;
+}
+
+#define CPU_FREQ DT_PROP(DT_PATH(cpus, cpu_0), clock_frequency)
+static int i2c_cc23x0_configure(const struct device *dev, uint32_t dev_config)
+{
+	const struct i2c_cc23x0_config *config = dev->config;
+	bool fast;
+
+	switch (I2C_SPEED_GET(dev_config)) {
+	case I2C_SPEED_STANDARD:
+		fast = false;
+		break;
+	case I2C_SPEED_FAST:
+		fast = true;
+		break;
+	default:
+		LOG_ERR("Unsupported speed");
+		return -EIO;
+	}
+
+	/* Support for slave mode has not been implemented */
+	if (!(dev_config & I2C_MODE_CONTROLLER)) {
+		LOG_ERR("Slave mode is not supported");
+		return -EIO;
+	}
+
+	/* This is deprecated and could be ignored in the future */
+	if (dev_config & I2C_ADDR_10_BITS) {
+		LOG_ERR("10-bit addressing mode is not supported");
+		return -EIO;
+	}
+
+	/* Enables and configures I2C master */
+	I2CControllerInitExpClk(config->base, fast);
+
+	CLKCTLEnable(CLKCTL_BASE, CLKCTL_I2C0);
+
+	return 0;
+}
+
+static void i2c_cc23x0_isr(const struct device *dev)
+{
+	const struct i2c_cc23x0_config *config = dev->config;
+	struct i2c_cc23x0_data *data = dev->data;
+	const uint32_t base = config->base;
+
+	if (I2CControllerIntStatus(base, true)) {
+		I2CControllerClearInt(base);
+
+		data->error = I2CControllerError(base);
+
+		k_sem_give(&data->complete);
```

Comment:
This results in a lot of back and forth to the thread which isn't free of charge, and will make it more complex to add any of the asynchronous APIs in the future. I would recommend changing the structure of the driver to accept a set of i2c_msgs and then work through them, starting the next transmit/receive directly from the ISR only giving complete when there no more messages to send.